speaker/bgp_controller.go

```diff
@@ -158,61 +138,258 @@ func (c *bgpController) ShouldAnnounce(l log.Logger, name string, svc *v1.Servic
 	return ""
 }
 
-// Called when either the peer list or node labels have changed,
-// implying that the set of running BGP sessions may need tweaking.
+// syncPeers updates the BGP peer list based on peers defined statically in the
+// configuration as well as any autodiscovered node peers.
 func (c *bgpController) syncPeers(l log.Logger) error {
-	var (
-		errs          int
-		needUpdateAds bool
-	)
-	for _, p := range c.peers {
+	if c.cfg == nil {
+		return nil
+	}
+
+	c.peersLock.Lock()
+	defer c.peersLock.Unlock()
+
+	// Merge static peers and discovered peers into a single slice representing
+	// the desired set of peers while ensuring we don't have two peers which
+	// would result in the same BGP session.
+	var wantPeers []*config.Peer
+	for _, p := range append(c.cfg.Peers, c.discoverNodePeers(l)...) {
```

Comment:
Would it make sense to set the discovered node peers as a controller property in `SetNode`?
It feels asymmetrical here (we set the peers on the cfg side, we set the node properties and evaluate the peers while syncing on the discovery side)

Reply:
Right now we're setting the node annotations and labels as controller properties in `SetNode` and return from `SetNode` without trying to discover node peers if the annotations/labels haven't changed since the last run.
With your suggestion, we would have to store a 3rd controller property - we can't discard `nodeAnnotations` and `nodeLabels` because we don't want to execute `discoverNodePeers` unnecessarily as it is a relatively expensive function.

I don't see the current state as "asymmetric": In `SetConfig` and `SetNode` we store configuration data (from the ConfigMap and from the node object, respectively) as controller properties. Then, in either function we only execute `syncPeers` if the configuration data has changed, and we do the "heavy lifting" inside `syncPeers`, i.e. we run the discover, remove duplicates etc.

WDYT?

Reply:
> Right now we're setting the node annotations and labels as controller properties in `SetNode` and return from `SetNode` without trying to discover node peers if the annotations/labels haven't changed since the last run.
> With your suggestion, we would have to store a 3rd controller property - we can't discard `nodeAnnotations` and `nodeLabels` because we don't want to execute `discoverNodePeers` unnecessarily as it is a relatively expensive function.
> 

This is a good point, did not think about the need of labels / annotations anyway to check if they differ.
Anyway, but being `discoverNodePeers` relatively expensive, we would save some cycles (if / when config changes more than the annotations) by doing that in setNode (which is kind of the same reason why we are caching labels / annotations).

> I don't see the current state as "asymmetric": In `SetConfig` and `SetNode` we store configuration data (from the ConfigMap and from the node object, respectively) as controller properties. Then, in either function we only execute `syncPeers` if the configuration data has changed, and we do the "heavy lifting" inside `syncPeers`, i.e. we run the discovery, remove duplicates etc.
> 

Re: asymmetric, we are setting the config peers (even because they don't need any manipulation) in `setConfig`, whereas we are re-evaluating them every time for the "node" side, so my feeling was that in `syncPeers` the "config peers" are already there while the "node peers" are being re-evaluated each time.

> WDYT?

Added some comments inline, I am also fine with the current implementation.